tools/ou3_p4_direct_word_contraction_certificate.py

```
@@ -0,0 +1,66 @@
+ if mode=='A' and proj:caps['bias_projection']=T._positive_root(B,L.mul_down(float(proj['interior_margin_lower_mps2']),sm))
+ sw=min(caps.values());W=L.mul_down(sw,sw);pf=L.add_up(1.,L.mul_up(B,sw));qp=L.div_up(L.mul_up(pf,sw),sm)
+ if not qp<=q or not qp<float(base['cayley_norm_limit']):raise RuntimeError('prefix')
+ if not L.mul_up(B,sw)<gap:raise RuntimeError('not strict')
+ return {'q':q,'B':B,'sqrtW':sw,'W':W,'qprefix':qp,'active_cap':min(caps,key=caps.get),'strict_gap':gap}
+def refine(mode,base,domain):
+ q0=float(base['thirdgen_selected_design_norm']);rows=[]
+ for q in T._grid(q0):
+  try:rows.append(candidate(mode,base,domain,q))
+  except Exception:pass
+ if not rows:raise RuntimeError('no direct contraction cell')
+ best=max(rows,key=lambda x:x['W']);before=float(base['certified_level_W']);m=copy.deepcopy(base);m.update({'direct_strict_word_contraction_search':True,'direct_contraction_candidates':rows,'direct_selected_design_norm':best['q'],'direct_selected_active_cap':best['active_cap'],'direct_strict_endpoint_gap_lower':best['strict_gap'],'certified_level_W_before_direct':before,'certified_level_W':max(before,best['W']),'certified_level_sqrt_W':max(float(base['certified_level_sqrt_W']),best['sqrtW']),'prefix_canonical_error_norm_upper':best['qprefix'],'direct_W_factor_lower':L.div_down(max(before,best['W']),before),'direct_claim':'W_end<W0 with positive strict gap; fixed delta/2 decrease is not claimed at the enlarged level'});return m
```

Ruff (0.16.2)

[error] Multiple statements on one line (semicolon)

(E702)